### PR TITLE
Computer takes a turn

### PR DESCRIPTION
**Pull request by thorncp**

### CODE REVIEW

File: lib/computer_turn.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+class ComputerTurn
+  attr_reader :grid, :view
+
+  def initialize(grid:, view:)
+    @grid = grid
+    @view = view
+  end
+
+  def take
+    view.render_computer_thinking
```
Comment (baccigalupi) [inline lib/computer_turn.rb:10]: Love this name!

File: lib/computer_turn.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    mark_random_cell
+  end
+
+  private
+
+  def mark_random_cell
+    grid.mark(coordinates: random_coordinates, value: "O")
```
Comment (baccigalupi) [inline lib/computer_turn.rb]: So, now that there are two (and only two) players. I am wondering about this value. Couldn't there be two methods `mark_player` and `mark_computer` so that this '0' and 'X' notation is encapsulated/hidden?

File: lib/computer_turn.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  end
+
+  def random_coordinates
+    grid.available_coordinates.sample
```
Comment (baccigalupi) [inline lib/computer_turn.rb]: Law of Demeter!!! What about asking the grid for a random coordinate?

File: lib/grid.rb [hunk 4]
```diff
@@ -1,10 +1,16 @@
+require_relative "coordinates"
+
 class Grid
   attr_reader :cells
 
   def initialize(cells = {})
     @cells = cells
   end
 
+  def available_coordinates
```
Comment (baccigalupi) [inline lib/grid.rb]: Originally, having the valid values over in the Coordinates made a lot of sense, but now that data seems much more related to the nature of the grid. What is and isn't in the grid ... is about the grid. That makes me wonder what kind of methods you need to answer the questions of the other classes.

Comment (baccigalupi) [reply lib/grid.rb]: Also, it seems like an encapsulation smell that we know over here that valid values is an array.

File: lib/text_view.rb [hunk 5]
```diff
@@ -16,4 +16,11 @@ def render_coordinates_error(coordinates)
 
     ERROR
   end
+
+  def render_computer_thinking
+    output.puts(<<~BRAINS)
```
Comment (baccigalupi) [inline lib/text_view.rb:21]: Love it!

File: spec/game_spec.rb [hunk 6]
```diff
@@ -2,6 +2,7 @@
 
 RSpec.describe Game do
   it "prints a grid" do
+    srand(0)
```
Comment (baccigalupi) [inline spec/game_spec.rb:5]: Shoot, I remember this (ish) from C. What does this do?

File: lib/text_view.rb (hunk 5 above)
Comment (baccigalupi) [reply lib/text_view.rb:21]: Maybe it should also sleep for a bit too! LOL. But that wouldn't belong here because ... this is a printer.

File: spec/grid_spec.rb [hunk 7]
```diff
@@ -1,6 +1,30 @@
 require_relative "../lib/grid"
 
 RSpec.describe Grid do
+  describe "#available_coordinates" do
+    it "returns coordinates that have not been used" do
+      grid = Grid.new("A1" => "X", "B2" => "O", "C3" => "X")
+
+      expect(grid.available_coordinates).to match_array %w[A2 A3 B1 B3 C1 C2]
```
Comment (baccigalupi) [inline spec/grid_spec.rb]: This test doesn't feel good either. The data escaping this way feels like it could lead to too much intimate knowledge/coupling.

**Review by baccigalupi — CHANGES_REQUESTED** [review]
Data encapsulation!! But this looks good otherwise!

File: lib/computer_turn.rb (hunk 2 above)
Comment (thorncp) [reply lib/computer_turn.rb]: Ooh, that's a good idea.

File: lib/text_view.rb (hunk 5 above)
Comment (thorncp) [reply lib/text_view.rb:21]: I definitely want to do this, but I also don't want the test suite to slow down because of it. I'm thinking of doing an addition at the end where I add a way to control the sleep time, so the tests can make it zero.

File: spec/game_spec.rb (hunk 6 above)
Comment (thorncp) [reply spec/game_spec.rb:5]: Seeds the random number generator, so the test can be predictable. I couldn't think of an alternative that didn't use mocking.

File: lib/computer_turn.rb [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+  def take
+    view.render_computer_thinking
+    mark_random_cell
+  end
+
+  private
+
+  def mark_random_cell
+    grid.mark_computer(coordinates: random_coordinates)
```
Comment (baccigalupi) [inline lib/computer_turn.rb:17]: Cool!

File: lib/computer_turn.rb [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+  end
+
+  def random_coordinates
+    grid.random_coordinates
```
Comment (baccigalupi) [inline lib/computer_turn.rb:21]: And this too!

File: lib/grid.rb [hunk 10]
```diff
@@ -1,15 +1,41 @@
...
+    %w[A1 A2 A3 B1 B2 B3 C1 C2 C3]
+  end
+
   def initialize(cells = {})
     @cells = cells
   end
 
-  def mark(coordinates:, value:)
-    cells[coordinates] = value
+  def mark_computer(coordinates:)
```
Comment (baccigalupi) [inline lib/grid.rb:12]: I love these a ton. They are the same shape. But also we don't need to know outside this class what the data looks like.

File: lib/text_view.rb (hunk 5 above)
Comment (baccigalupi) [reply lib/text_view.rb:21]: Oh, yeah!

File: spec/coordinates_spec.rb [hunk 11]
```diff
@@ -2,32 +2,24 @@
...
-    it "is false when in a column before the first column in the grid" do
-      expect(Coordinates.new(value: "C0")).not_to be_valid
-    end
+    it "is false when the grid validates it to be" do
+      grid = double(valid_coordinates?: false)
+      coordinates = Coordinates.new(value: "A1", grid: grid)
 
-    it "is false when in a column after the last column of the grid" do
-      expect(Coordinates.new(value: "B4")).not_to be_valid
+      expect(coordinates).not_to be_valid
```
Comment (baccigalupi) [inline spec/coordinates_spec.rb:16]: Nice on these two tests!

File: spec/game_spec.rb (hunk 6 above)
Comment (baccigalupi) [reply spec/game_spec.rb:5]: Oh, cool!

File: spec/grid_spec.rb [hunk 12]
```diff
@@ -1,16 +1,75 @@
...
+        "B1" => "O",
+        "B2" => "O",
+        "B3" => "O",
+        "C1" => "X",
+        "C2" => "X",
+        "C3" => "X"
+      )
+
+      expect(grid.random_coordinates).to be_nil
+    end
```
Comment (baccigalupi) [inline spec/grid_spec.rb:46]: Nice spec!

File: spec/grid_spec.rb [hunk 13]
```diff
@@ -1,16 +1,75 @@
...
+  end
+
+  describe "#valid_coordinates?" do
+    it "is true when the value is in the grid" do
+      expect(Grid.new.valid_coordinates?("A1")).to be true
+      expect(Grid.new.valid_coordinates?("B2")).to be true
+      expect(Grid.new.valid_coordinates?("C3")).to be true
+    end
+
+    it "is false when in a row before the first row of the grid" do
```
Comment (baccigalupi) [inline spec/grid_spec.rb]: Didn't notice these before, but the descriptions don't make a lot of sense. 

This one looks like `is false when the the coordinates don't match the pattern`

File: spec/grid_spec.rb [hunk 14]
```diff
@@ -1,16 +1,75 @@
...
+      expect(Grid.new.valid_coordinates?("@1")).to be false
+    end
+
+    it "is false when in a row after the last row in the grid" do
+      expect(Grid.new.valid_coordinates?("C0")).to be false
+    end
+
+    it "is false when in a column before the first column in the grid" do
+      expect(Grid.new.valid_coordinates?("C0")).to be false
```
Comment (baccigalupi) [inline spec/grid_spec.rb:61]: These two have the same value as input.

**Review by baccigalupi — APPROVED** [review]
Looks good. Just some minor stuff.

File: spec/grid_spec.rb (hunk 13 above)
Comment (thorncp) [reply spec/grid_spec.rb]: Yeah, it's trying to be symmetrical and cover the case before the rows start. I picked `@` because it's the ASCII character before `A`, but I agree it doesn't really match the description. Updating this to be about format, and not having a spec for "before," seems appropriate.
